### PR TITLE
feat: add group object 

### LINKED ISSUE #125: [users] Implementing user_group
Following up on #45, I think it is sound to have a discussion about the multiple access levels in the API. With the current design of scopes, we cannot efficiently separate accesses of large user groups.

## Requirements
- We need to restrict "superadmin" from common users
- Users need to be segmented into groups, where each group only has access to a restricted scope (i.e. a local firefighter, while having a "user" scope, should not have access to the same data as someone from another region, while that person is also a "user")
- Groups need to be easily editable
- security on routes will have to combine those to provide proper and secure access to each route

## Design suggestion
Simply put, I believe two fields are required:
- access_type: admin, user, device, me
- access_group: ID that would reference a group (we could add a `groups` table with `group_id`, and `group_name`)

### PR DESCRIPTION
**Pull request by florianriche**
Create a new object called "group".
That object is linked to a user entity.
Right now, only the Unit tests and the code related to the group management has been added.

Before implementing the code linked to the authorization management, I would like opinions on this schema.

Closes #125

### CODE REVIEW

**Comment by frgfm** [comment]
Thanks for the PR @florianriche !
Since this is a rather big deal as a feature, I think we will most likely have to split this into several PRs depending on the design :sweat_smile:
Let's discuss this in the issue #125 quickly :ok_hand:

**Comment by florianriche** [comment]
> Thanks for the PR @florianriche !
> Since this is a rather big deal as a feature, I think we will most likely have to split this into several PRs depending on the design 
> Let's discuss this in the issue #125 quickly

I have updated the PR to remove the reference to group_id in user table

File: src/tests/routes/test_alerts.py [hunk 1]
```diff
@@ -31,6 +31,10 @@
     {"id": 2, "login": "connected_user", "access_id": 2, "created_at": "2020-11-13T08:18:45.447773"},
 ]
 
+GROUP_TABLE = [
+    {"id": 1, "name": "test_group"},
+]
+
```
Comment (frgfm) [inline src/tests/routes/test_alerts.py:37]: Let's remove everything in the unittests that doesn't involve groups yet (so everything apart from test_groups.py)

File: src/tests/routes/test_devices.py [hunk 2]
```diff
@@ -17,6 +17,10 @@
     {"id": 2, "login": "connected_user", "access_id": 2, "created_at": "2020-11-13T08:18:45.447773"},
 ]
 
+GROUP_TABLE = [
+    {"id": 1, "name": "test_group"},
+]
+
```
Comment (frgfm) [inline src/tests/routes/test_devices.py:23]: same

File: src/app/api/schemas.py [hunk 3]
```diff
@@ -74,6 +73,19 @@ class UserCreation(UserInfo):
     access_id: int = Field(..., gt=0)
 
 
+# Groups
+class GroupBase(BaseModel):
```
Comment (frgfm) [inline src/app/api/schemas.py:77]: You can rename this into GroupIn, no need for GroupBase

File: src/app/api/schemas.py [hunk 4]
```diff
@@ -74,6 +73,19 @@ class UserCreation(UserInfo):
...
+    name: str = Field(..., min_length=3, max_length=50, example="Fireman85")
+
+
+class GroupIn(GroupBase):
+    pass
```
Comment (frgfm) [inline src/app/api/schemas.py:82]: we can remove this

File: src/app/api/schemas.py [hunk 5]
```diff
@@ -74,6 +73,19 @@ class UserCreation(UserInfo):
...
+
+
+class GroupOut(GroupBase, _Id):
```
Comment (frgfm) [inline src/app/api/schemas.py:85]:
```
class GroupOut(GroupIn, _Id):
```

**Review by frgfm — COMMENTED** [review]
Hey there!
Thanks for the PR, I added a few comments. I suggest waiting for #138 to be merged before writing the actual group unittests

**Review by frgfm — CHANGES_REQUESTED** [review]
Just need to fix flake8 checks on the new unittests ;)

**Comment by frgfm** [comment]
Shoot, this also needs some merging @florianriche :sweat_smile:

File: src/tests/routes/test_groups.py [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+]
+
+
+def compare_entries(ref, test):
+    for k, v in ref.items():
+        if isinstance(v, float):
+            # For float issues
+            assert abs(v - test[k]) < 1E-5
+        else:
+            assert v == test[k]
```
Comment (frgfm) [inline src/tests/routes/test_groups.py:34]: Considering that there is no float field in the group table, perhaps we don't need this?

File: src/tests/routes/test_groups.py [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+async def test_get_group(test_app_asyncio, init_test_db, group_id, status_code, status_details):
+
+    response = await test_app_asyncio.get(f"/groups/{group_id}")
+    response_json = response.json()
+    assert response.status_code == status_code
+
+    if isinstance(status_details, str):
+        assert response.json()['detail'] == status_details
+    if response.status_code == 200:
+        compare_entries(response_json, GROUP_TABLE[group_id - 1])
```
Comment (frgfm) [inline src/tests/routes/test_groups.py:65]: I think we can check for perfect match (no float field here)

File: src/tests/routes/test_groups.py [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+
+
+@pytest.mark.asyncio
+async def test_fetch_groups(test_app_asyncio, init_test_db):
+
+    response = await test_app_asyncio.get("/groups/")
+    assert response.status_code == 200
+    response_json = response.json()
+    for (i, entry) in enumerate(response_json):
+        compare_entries(entry, GROUP_TABLE[i])
```
Comment (frgfm) [inline src/tests/routes/test_groups.py:75]: same here

File: src/tests/routes/test_groups.py [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+    auth = await pytest.get_token(ACCESS_TABLE[access_idx]['id'], ACCESS_TABLE[access_idx]['scopes'].split())
+
+    response = await test_app_asyncio.delete(f"/groups/{group_id}/", headers=auth)
+    assert response.status_code == status_code
+
+    if isinstance(status_details, str):
+        assert response.json()['detail'] == status_details
+
+    if response.status_code // 100 == 2:
+        compare_entries(response.json(), GROUP_TABLE[group_id - 1])
```
Comment (frgfm) [inline src/tests/routes/test_groups.py:167]: same here

**Review by frgfm — APPROVED** [review]
Thanks for the PR! I only added a last comment about `compare_entries`: since there is no float field in the table, we can check for exact match without it. But there the rest looks good to me!

**Pull request merged**